File: 11-agenci/solution.py
from collections import defaultdict, deque

def bfs(graph, a1, a2):
    queue_after_agent2_move = deque([(a1, a2)])

    visited_after_agent1_move = set()
    visited_after_agent2_move = set([(a1, a2)])

    days = 0
    while queue_after_agent2_move:
        queue_after_agent1_move = deque()

        #ruch agenta1
        while queue_after_agent2_move:
            agent1, agent2 = queue_after_agent2_move.popleft()
            if agent1 == agent2:
                return days
            
            # przechodzimy po sasiednich miastach agenta1
            # dodajemy nowe stany po przemieszczeniu sie agenta1
            for new_agent1 in graph[agent1]:
                state = (new_agent1, agent2)
                if not state in visited_after_agent1_move:
                    visited_after_agent1_move.add(state)
                    queue_after_agent1_move.append(state)
        
        # ruch agenta2
        while queue_after_agent1_move:
            agent1, agent2 = queue_after_agent1_move.popleft()

            # przechodzimy po sasiednich miastach agenta2
            # dodajemy nowe stany po przemieszczeniu sie agenta2
            for new_agent2 in graph[agent2]:
                state = (agent1, new_agent2)
                if not state in visited_after_agent2_move:
                    visited_after_agent2_move.add(state)
                    queue_after_agent2_move.append(state)

        days += 1
    return 'NIE'

n, m = map(int, input().split())
agent1, agent2 = map(int, input().split())

graph = defaultdict(list)
for _ in range(m):
    a, b = map(int, input().split())
    graph[a].append(b)

res = bfs(graph, agent1, agent2)
print(res)

File: 11-agenci/solution.cpp
#include <iostream>
#include <queue>
#include <map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

struct PairHash {
    template <class T1, class T2>
    size_t operator() (const pair<T1, T2> &v) const {
        auto [a, b] = v;
        // wydajne obliczanie hasha pary intow
        size_t h = (a*a + 3*a + 2*a*b + b + b*b) / 2;
        return h;
    }
};

using Graph = map<int, vector<int>>;
using VisitedSet = unordered_set<pair<int, int>, PairHash>;
using Queue = queue<pair<int, int>>;


bool was_visited(VisitedSet &visited, pair<int, int> state) {
    return visited.find(state) != visited.end();
}

pair<int, int> queue_pop(Queue &queue) {
    pair<int, int> popped_element = queue.front();
    queue.pop();
    return popped_element;
}

int bfs(Graph &graph, int a1, int a2) {
    Queue queue_after_agent2_move;
    queue_after_agent2_move.push({a1, a2});

    unordered_set<pair<int, int>, PairHash> visited_after_agent1_move;
    unordered_set<pair<int, int>, PairHash> visited_after_agent2_move;
    visited_after_agent2_move.insert({a1, a2});

    int days = 0;
    while (!queue_after_agent2_move.empty()) {
        Queue queue_after_agent1_move;

        // ruch agenta1
        while (!queue_after_agent2_move.empty()) {
            auto [agent1, agent2] = queue_pop(queue_after_agent2_move);
            
            if (agent1 == agent2)
                return days;

            // przechodzimy po sasiednich miastach agenta1
            // dodajemy nowe stany po przemieszczeniu sie agenta1
            for (int new_agent1 : graph[agent1]) {
                pair<int, int> state = make_pair(new_agent1, agent2);
                if (!was_visited(visited_after_agent1_move, state)) {
                    visited_after_agent1_move.insert(state);
                    queue_after_agent1_move.push(state);
                }
            }
        }

        // ruch agenta2
        while (!queue_after_agent1_move.empty()) {
            auto [agent1, agent2] = queue_pop(queue_after_agent1_move);

            // przechodzimy po sasiednich miastach agenta2
            // dodajemy nowe stany po przemieszczeniu sie agenta2
            for (int new_agent2 : graph[agent2]) {
                pair<int, int> state = make_pair(agent1, new_agent2);
                if (!was_visited(visited_after_agent2_move, state)) {
                    visited_after_agent2_move.insert(state);
                    queue_after_agent2_move.push(state);
                }
            }
        }      
        days++;
    }

    return -1;
}


int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int n, m;
    cin >> n >> m;

    int agent1, agent2;
    cin >> agent1 >> agent2;

    Graph graph;
    for (int i = 0; i < m; i++) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
    }

    int days = bfs(graph, agent1, agent2);

    if (days == -1)  
        cout << "NIE";
    else            
        cout << days;
    return 0;
}